### ISSUE TITLE
Resolve config specific imports for Dart 2

### ISSUE THREAD

**Issue by leafpetersen** [dart-lang/sdk#32960]
Config specific imports are only partially supported by Dart tools (notably the build tools).  There are some limited existing uses of config specific imports, mostly for testing purposes.  There are possibly planned uses in the future for Dart SASS.  

This issue is for discussion of possible short term resolutions for Dart 2 to avoid shipping something that we will later regret.  Options on the table:
  - Fully unship the feature
    - pro: the feature as it exists is problematic
    - con: it fulfills a serious need
  - Ship the feature as it is
    - pro: the feature is useful
    - con: many build tools do not and may not ever support it
  - Move the feature back behind a flag a-la super-mixins with the intent of revisiting the feature in the next year (possibly with breaking changes)
    - pro: gives us license to make breaking changes
    - con: leaves the feature in an unfortunate heisen-state, users can't rely on it

@matanlurey @sigmundch Can you summarize on the issue why build tools can't support this?

@nex3 Can you comment on the Dart SASS requirements?

cc @zanderso @lrhn @anders-sandholm @vsmenon @munificent @a-siva

**Comment by leafpetersen** [comment]
Summarizing some discussion so far from elsewhere:

  - Dart currently assumes that platform defines may only be available at runtime
  - In order to do modular compilation with config specific imports, we need to know the platform defines
    - Unless we define an interface conformance check so that we can compile against default interfaces, which isn't happening any time soon

So questions I have are:
  - Why is it infeasible to allow modular/AOT compilers to require a specific set of platform defines at compile time (with an implied link time check for consistency)?
    - We could still *allow* platform defines to be provided at runtime for tools that support it (e.g. the vm).
  - Are there existing uses of platform defines which are incompatible with the above?
  - If so, could we not make config specific imports use something other than general platform defines?

**Comment by nex3** [comment]
I have an in-progress document that defines semantics for interface conformance that's more thorough than the sketch that's in the original proposal, and that I hope will be easier to implement as well. If there's interest, I could flesh that out and send it around.

> Can you comment on the Dart SASS requirements?

Dart Sass needs to be able to target three platforms with the same codebase: the Dart VM, Node, and browsers. Right now, we look at the `dart.library.io` and `dart.library.js` automatic variables to do so.

**Comment by sigmundch** [comment]
> Can you summarize on the issue why build tools can't support this?

It comes down to the modularity points you made earlier: without conformance, build systems don't have a clear design:
  * they could create build artifacts for different configurations, but that doesn't scale when combining configurations from dependencies (configuration explosion); or
  * they could otherwise store configuration-specific information on each build artifact and validate consistency at link time, but that makes maintenance of large codebases really hard for developers. That is because issues may arise down the line when someone happens to import two packages that have inconsistent platform defines in their configurations.

I want to add another option on the table:

* ship the feature that is implemented in the CFE
  * pros: it is useful, it is implemented, and build tools can support this.
  * cons: it means making a last minute spec change (although small).

The CFE implementation has one extra limitation: only platform defines that are invariant in a platform are visible for config specific imports. That is, you can see only defines of the from `dart.library.foo` which indicate whether `dart:foo` is supported by the platform.

Like Dart-SASS, we believe most uses of config-specific imports are choosing between platforms, and hence only need this limited form. Note: dart2js defines a different platform when it compiles for the browser or for compiling for node (e.g. `dart:html` is only supported in the former)

Build systems can support this because they now only need to create a fixed number of build artifacts: one per platform.

**Comment by leafpetersen** [comment]
> ship the feature that is implemented in the CFE

@zanderso @nex3 Presumably this covers both of your existing use cases?

@vsmenon Had one similar/alternative suggestion, which is that we could provide a way to redirect `dart:foo` libraries.  So `dart:html` points to the real `dart:html` on some platforms and to an either platform defined or possibly even user defined shim library on others.   Would this cover existing use cases?

**Comment by leafpetersen** [comment]
cc @grouma for comment on how much work it would be for the build system to support:
  - locked down per platform configurations
  - threading arbitrary build configurations through

**Comment by leafpetersen** [comment]
> I have an in-progress document that defines semantics for interface conformance that's more thorough than the sketch that's in the original proposal, and that I hope will be easier to implement as well. If there's interest, I could flesh that out and send it around.

I'm personally quite interested in doing something around this at some point, but there's no way we can get it for Dart 2, and seems unlikely for the next release or two.  But I'd definitely be happy to see work done on this to work towards some future release.

**Comment by leafpetersen** [comment]
> Put the feature, as implemented in the CFE, behind an experiment to unblock current users, but not block us from refining or changing the feature in the future

This reminds me of the other question: @nex3 for Dart SASS do you need to ship a package that relies on this feature, or do you just need this for building Dart SASS itself?  That is, if we move this behind a flag, will we need to worry about end users having to set the flag in order to use important packages, or can we firewall this into code used to build tools?

**Comment by nex3** [comment]
> @vsmenon Had one similar/alternative suggestion, which is that we could provide a way to redirect dart:foo libraries. So dart:html points to the real dart:html on some platforms and to an either platform defined or possibly even user defined shim library on others. Would this cover existing use cases?

This is pretty similar to my original proposal years ago, which was to provide stub libraries that just throw (like we currently do for `dart:io`). I like the idea in general, but I'm worried that getting a good design in place and implemented by Dart 2 is infeasible.

> This reminds me of the other question: @nex3 for Dart SASS do you need to ship a package that relies on this feature, or do you just need this for building Dart SASS itself? That is, if we move this behind a flag, will we need to worry about end users having to set the flag in order to use important packages, or can we firewall this into code used to build tools?

The `sass` package on pub.dartlang.org does use configuration-specific imports; they aren't just build-time, because there's no build before publishing a VM package. So yes, it would be a pretty serious usability problem if we had to make downstream users of Sass in Dart toggle this flag. *However*, I don't think people using Sass from Dart that's then compiled to JS is a use-case we care a lot about

This suggests a possible solution: we could support the CFE subset, and put it behind a flag *only for dart2js and DDC*. The VM would be able to consume config-specific imports by default, but anyone using it in practice would have to opt into the experiment somewhere.

**Comment by grouma** [comment]
Discussed offline with Leaf to provide some context of the current platform support which is related to config specific imports.

Some notes:
- In our build rules we have the concept of platforms: https://github.com/dart-lang/rules_dart/blob/master/dart/build_rules/common/context.bzl#L55
- Currently this is used only internally to ensure compatible platforms between dart libraries
- In the near future we intend to make the internal analyzer platform aware, meaning we can adjust the SDK given the platform of the dart library
- Besides analysis, we do not change the behavior of builds based on the platform and have no current plans to do so
- As far as I know package:build has no concept of platforms and there is no work planned to provide platform support
- The pub site has basic platform support by showing the inferred support for a package. It uses https://pub.dartlang.org/packages/pana to do analysis.

**Comment by natebosch** [comment]
I'm in favor of keeping support hidden behind a flag so that we can delay the decision on what the final feature will look like. I think it's plausible we can support the "platform" specific import idea, but I'd really like to have at least a prototype working in build_runner, bazel, and flutter, before we decide to ship it as a fully supported feature.

**Comment by zanderso** [comment]
> There are some limited existing uses of config specific imports, mostly for testing purposes.

To clarify a bit, the uses in Fuchsia are limited, but the alternatives to using them involve writing one or more new embeddings of the VM or doing a large refactoring of FIDL bindings generation.

> @zanderso @nex3 Presumably this covers both of your existing use cases?

The current implementation in the CFE is working great for the Fuchsia use-case.

I would like to hear more details about how this feature would work when under a flag. What would accept the flag, and what would it mean?

**Comment by leafpetersen** [comment]
> I would like to hear more details about how this feature would work when under a flag. What would accept the flag, and what would it mean?

I think the proposal would be for a flag like the current `--super-mixins`.  That is, without passing `--config-specific-imports`, the syntax would be rejected, with the flag it would be accepted.  The flag would only be supported on the VM (and possibly native AOT) side.  So Fuschia would have to be set up to run both natively and in test mode with the flag enabled.  Presumably it would be easy to set the native embedder up to just turn the flag on by default (much as Flutter does with `--super-mixins`).  So the question is whether it would be easy to do this on the host side as well.

**Comment by zanderso** [comment]
By "supported on the VM" do you mean "supported by the CFE when targeting the VM" ?

**Comment by jakemac53** [comment]
Additional side note here, but wouldn't any constant which is defined expliclity in Dart code also fall into the "only platform defines that are invariant in a platform" bucket?

This would effectively change the proposal to "don't expose environment constants to config specific imports" (and maybe some others?) instead of locking it down to an explicit whitelist of constants from the sdk.

The reason this makes a difference is it allows the ability for build systems to generate different Dart files based upon configuration, and it leaves it up to the build system itself to appropriately invalidate all modules that depend on those generated dart files.

This has several benefits over environment variables as well, since it makes the dependency on these constants explicit instead of implicit, allowing build systems to do smarter invalidation.

At the same time, it provides most of the same benefits I believe?

See https://github.com/dart-lang/build/issues/1053 for some context, we have users who want to expose `-D` vars for `build_runner`, but doing so would cause us to invalidate the entire world whenever any of those change.

**Comment by leafpetersen** [comment]
> By "supported on the VM" do you mean "supported by the CFE when targeting the VM" ?

I think so yes.  Terminology gets confusing here.

**Comment by leafpetersen** [comment]
> Additional side note here, but wouldn't any constant which is defined expliclity in Dart code also fall into the "only platform defines that are invariant in a platform" bucket?

I'm not sure I follow where you're going.  Are you suggesting to move this over to the build system?  That is, have the build system conditionally replace (e.g.) "platform_defines.dart" with "platform_defines_vm.dart" or "platform_defines_dart2js.dart" as appropriate, where those files define the appropriate const values?

**Comment by leafpetersen** [comment]
Ok, here's my proposal.

TLDR: let's do what @sigmundch suggests above and commit to supporting what the new front end currently implements, without a flag. 

> The CFE implementation has one extra limitation: only platform defines that are invariant in a platform are visible for config specific imports. That is, you can see only defines of the from dart.library.foo which indicate whether dart:foo is supported by the platform.

Rationale:
  - I am highly skeptical that we will ever be able to share compilation artifacts between VM and JS platforms
  - Within JS platforms, the compilation artifacts should be as sharable with this feature as without
  - For analysis, I think we can live with the restriction to having to choose a platform to analyze against if we want to share compilation artifacts between build and analysis steps
  - This supports all of our known important use cases
  - I don't think this prevents us from implementing better config specific imports
    - the syntactic limitation to `dart.library.foo` means that we are reserving only that limited syntax

@lrhn WDYT?

@munificent @sigmundch @vsmenon @bwilkerson @nex3 @zanderso @matanlurey please let me know if you have strong objections to taking this approach.

**Comment by lrhn** [comment]
On Wed, May 9, 2018 at 11:41 PM Leaf Petersen <notifications@github.com>
wrote:

> Ok, here's my proposal.
>
> TLDR: let's do what @sigmundch <https://github.com/sigmundch> suggests
> above and commit to supporting what the new front end currently implements,
> without a flag.
>
> The CFE implementation has one extra limitation: only platform defines
> that are invariant in a platform are visible for config specific imports.
> That is, you can see only defines of the from dart.library.foo which
> indicate whether dart:foo is supported by the platform.
>
> Rationale:
>
>    - I am highly skeptical that we will ever be able to share compilation
>    artifacts between VM and JS platforms
>    - Within JS platforms, the compilation artifacts should be as sharable
>    with this feature as without
>    - For analysis, I think we can live with the restriction to having to
>    choose a platform to analyze against if we want to share compilation
>    artifacts between build and analysis steps
>    - This supports all of our known important use cases
>    - I don't think this prevents us from implementing better config
>    specific imports
>       - the syntactic limitation to dart.library.foo means that we are
>       reserving only that limited syntax
>
> @lrhn <https://github.com/lrhn> WDYT?
>

Since we already have the feature, I guess this is a good trade-off.
I wouldn't add a feature with that design from scratch, though - it's too
limited in scope to deserve having such a complicated syntax.

 So, LGTM.

@munificent <https://github.com/munificent> @sigmundch
> <https://github.com/sigmundch> @vsmenon <https://github.com/vsmenon>
> @bwilkerson <https://github.com/bwilkerson> @nex3
> <https://github.com/nex3> @zanderso <https://github.com/zanderso>
> @matanlurey <https://github.com/matanlurey> please let me know if you
> have strong objections to taking this approach.
>
> —
>

-- 
Lasse R.H. Nielsen - lrn@google.com
'Faith without judgement merely degrades the spirit divine'
Google Denmark ApS - Frederiksborggade 20B, 1 sal - 1360 København K
- Denmark - CVR nr. 28 86 69 84

**Comment by matanlurey** [comment]
I'm admittedly not _excited_, but this seems OK, with caveats that we still won't be able to support this syntax in Build Runner or Bazel as-is, at least not as specified. If that is OK, then LGTM. Otherwise we might want to reconsider the flag approach.

**Comment by natebosch** [comment]
> Since we already have the feature

Sort of... dart2js and the VM support it when you use them directly. The way most of our users (will) build with dart2js is build_runner and that doesn't support it. The way our internal users build for either dart2js or VM is bazel and that doesn't support it.

Given that running VM tests is going to become painfully slow in Dart 2 the build system is likely going to end up in the dev cycle for people outside of our angular users. My fear is that we say today "We're ok not supporting this in build_runner" and then end up with a P0 to support it some time down the line.

**Comment by leafpetersen** [comment]
> The way most of our users (will) build with dart2js is build_runner and that doesn't support it. The way our internal users build for either dart2js or VM is bazel and that doesn't support it.

I don't really understand this.  Are you claiming that I can, with build_runner or bazel, do the following workflow:
  - build massive_app
  - run_vm massive_app // Doesn't rebuild
  - run_dart2js massive_app // Still doesn't rebuild

?

My understanding is not.  My understanding of the workflow is more like:
  - build_for_vm massive_app
  - build_for_dart2js massive_app

If the former is in fact the intended workflow then I agree that we don't support this feature, but I also don't understand how that works - VM and dart2js don't even share the same sdk.

If the latter is the intended workflow then I think that build_runner and bazel already support it just fine.

**Comment by leafpetersen** [comment]
> the workflow we want is

I am *highly* skeptical that you can get what you want here, and am very concerned if we are trying to build our ecosystem around the assumption that we can.  A few implications of this:
  - kernel can do *essentially no constant evaluation* (numeric results depend on platform)
  - the SDK libraries must be written to guarantee that implementation types *never* leak, even through inference (since VM and web have different implementation types)
  - kernel can do very little (essentially no) pre-optimization
    - inlining depends on what to inline, which is platform specific
    - type analysis propagates implementation types which is a no-no

What are the use cases where we want to have one build shared between the VM and dart2js?

**Comment by sigmundch** [comment]
I share @leafpetersen concerns. Things may have changed, but I believe long ago we were planning to create a modular tool chain in bazel by keeping assets for each platform separate. In essence, we would:
* build modular kernel files per platform (web, vm)
* feed these modular inputs to the appropriate target (vm/dart2js)

More often than not, users have a code-flow mostly in one platform, so it's possible that the impact of duplicating work per platform is low.

**Comment by natebosch** [comment]
> What are the use cases where we want to have one build shared between the VM and dart2js?

For performance reasons we'd like to share as much of the work as possible but it isn't strictly required. We'd be fine duplicating work _n_ times where _n_ is a fixed number of platforms that the build system knows about. I think what's important in that case from a build system perspective is:

- What is _n_
- Will we be able to hardcode knowledge of the variants in the build systems
- **Are we scoping out the work to do this in the build systems**

It's less important for us that we have sharing across platforms than that any given platform can be compiled in a modular way. So far DDC is our only platform that we're targeting with modular builds but we have short term plans to try to get module builds working for VM, and I think we've expected we'll be able to feed kernel files to dart2js to get some notion of incremental builds there as well.

If we can't get this working in the build systems and we're forced with a monolithic from-source VM run if any transitively imported code runs that adds a lot more pressure on other packages like package:test to add complexity in how it spawns isolates in order to pay a high cost for every test file in a package. https://github.com/dart-lang/sdk/issues/32953#issuecomment-383758520

> Things may have changed, but I believe long ago we were planning to create a modular tool chain in bazel by keeping assets for each platform separate.

It isn't on any roadmap that I'm aware of. If we're OK forcing it to be then I think this feature is fine assuming we have some time to gain confidence in our ability to support it that way.

@leafpetersen and I will chat about this soon on a higher bandwidth channel than github issues.

**Comment by leafpetersen** [comment]
@natebosch Understood.  

To jump up a level and give some context, there's a question of whether Dart should try to be a platform-independent language, or a cross-platform language.  I am fairly strongly of the opinion that it should be the latter.  We've tried very hard to pretend that we're the former, at the cost of trying desperately to pretend that dart2js is just "implementing Dart wrong" rather than "implementing Dart right for the platform it's targeting".  I really think we will better serve our users by accepting that platforms are a leaky abstraction, and giving them the tools to plug the leaks.  I think that does mean accepting that "platform" is going to have to be a concept in the ecosystem.

I definitely respect the worries about biting off unscoped work, and I'm ok pushing off first class support for this.  But I'm very worried about baking the concept of platform-independence into our tools now, because I think we're likely to just have to revisit that later, potentially at significant cost.  This seems to me to be pretty close to the minimal feature, and it feels like it's within our reach to support it.  I'm ok with making it more minimal if that helps:  for example we could say that there are only three platforms for now (platform.web, platform.fuschia, platform.io).  It doesn't feel like that makes things substantially simpler though.

**Comment by jakemac53** [comment]
There is another which I have here concern here, which is that summary files today (and kernel files later) are already being used to generate _cross platform dart code_ (internally at least, but externally eventually).

Some existing examples include json serialization (package:json_serializable, package:dson, etc) and dependency injection (package:inject). In practice, I doubt either of these would actually be affected by subtle differences in sdk summaries, but it would be possible.

Today, we are using the dev compiler sdk summary to do this, for all dart codegen. This has been a known issue for some time but hasn't been a real problem in practice yet.

In both bazel and build_runner today we do not have the ability to have multiple versions of a single dart file (ex: one per platform), so we have to choose some canonical sdk summary file to use. Reworking this now to add support for different generated files per platform would be a significant amount of work in both bazel, build_runner, and possibly all the core tools that need to resolve uris as well.

> What are the use cases where we want to have one build shared between the VM and dart2js?

Tests are a very common scenario where this comes up. It is a common pattern to run unit tests on the vm and only web specific tests via dart2js/dartdevc.

Assuming that we determine it is acceptable/required to double the number of kernel files generated in this scenario - we still _cannot_ generate different dart files for each platform today. That means the only option as it exists today would be to require completely separate builds with a global flag determining the platform you are building for. This is essentially not possible to do on Tap internally (you have to get explicit permission to run tests with certain flags), and manually passing the right platform flag when running tests manually would be a non-starter I think, although we could possibly do smart tooling.

**Comment by jakemac53** [comment]
> we still cannot generate different dart files for each platform today.

Actually, let me clarify - we cannot generate different dart files with the same URI.

After talking out loud about this a bit I did realize one potentially workable solution, which is that any Builder that knows it needs to generate platform specific dart code would opt specifically into doing that, and produce dart files with different file paths (`.<platform>.dart`?).

We could create a separate api for doing exactly this, a `PlatformSpecificBuilder` which is invoked once _per platform_ and is given an analysis context that is specific to that platform for each invocation.

It would then be up to consumers of those files to use config specific imports (or explicit ones if they only support one platform) to import the right file.

This would still require some level of support in bazel and build_runner, but it fits in relatively well with existing concepts.

It doesn't fix the performance implications of doing # platforms times as much work though.

**Comment by leafpetersen** [comment]
> Tests are a very common scenario where this comes up. It is a common pattern to run unit tests on the vm and only web specific tests via dart2js/dartdevc.

Really?  I thought essentially all tests were on DDC now?

In any case, is it really common that you want to run both interactively?  My question is not whether we ever need to run tests on VM and on DDC, it's why we need to *share a build*.

Let me go back to the root cause of my confusion.  There exist 32 bit and 64 bit builds of C++ code (and ARM builds, etc).  There surely exist tests for that code.  Those tests surely rely on shared libraries.  Those tests, by definition, cannot be sharing build artifacts.   So I can't understand how it can be a limitation of bazel that stops us from doing this.  Bazel simply *must* support platform specific builds.  It's certainly true that two builds for different platforms don't share compilation artifacts.  Why do we care?

**Comment by leafpetersen** [comment]
> It is still (at least with Dart 1) a lot faster to run vm tests so afaik it is still common (maybe less so internally?).

There are essentially no Dart 1 tests internally, and other than flutter, no way to run Dart 2 tests with the VM, right?  Externally, there's essentially no way to run Dart 2 tests at all, I think. 

> There is a significant usability improvement of being able to run all your tests with a single invocation regardless of platform

Do C++ developers really have to do a different invocation for every platform they want to test on?  That seems unlikely to me.

**Comment by jakemac53** [comment]
> Do C++ developers really have to do a different invocation for every platform they want to test on? That seems unlikely to me.

I don't know how c++ handles it specifically - but again I am not saying its not possible to support this with bazel. What I am saying is its not possible with the current Dart support for bazel because we can't generate two files at the same path.

There are multiple options for how it _could_ be supported in the future, but this isn't something that was on the radar previously so somebody would have to formulate a plan and execute on it.

The same problem exists for build_runner as well - although the solution would likely look a lot different.

**Comment by leafpetersen** [comment]
> There are multiple options for how it could be supported in the future, but this isn't something that was on the radar previously so somebody would have to formulate a plan and execute on it.

Fair enough.  I think I'm of the opinion we need to do this.

**Comment by sigmundch** [comment]
@jakemac53 - good point about generated source files. I wanted to note that C++ vs Dart may have different tools at hand: bazel has global configurations (flags passed to bazel itself) that can be used to choose the target architecture in C++ builds (e.g. cpu=k8). Those global configurations create a parallel build-tree in bazel itself and essentially let you generate sources and artifacts that could be different without conflicts and transparently, they however don't let you mix and match artifacts built for different architectures (IIUC). Our support for Dart in bazel is done via skylark rules where all of our platforms/configurations are available at once on the same build-tree. We can make a bigger built-tree to keep artifacts separate internally, but when it comes to generating source files, that abstraction leaks: to import a generated Dart source-file you'd need to be aware of the platform.

A question I want to discuss is about the role of summaries in this story. In particular, I agree with @leafpetersen that we want to be a `cross-platform` language, and not `platform-independent`, but what about summaries? Can we aim to have `platform-independent` summaries for any given library and sdk?

@natebosch and @jakemac53 - you both make very good points, and I'd be interested in joining you to discuss more details about this when we are all back in one place next week.

**Comment by leafpetersen** [comment]
> Can we aim to have platform-independent summaries for any given library and sdk?

I really don't think that's compatible with this feature, and I'm not convinced it's even compatible with the our SDK.  As soon as a platform specific type leaks out of the SDK, all of your inferred types from then are potentially bogus.  If we're very very careful about where we allow patch files, maybe we can avoid this, I don't know.

**Comment by leafpetersen** [comment]
> Can we aim to have platform-independent summaries for any given library and sdk?

I again have to ask though, what's the motivation for wanting this?

**Comment by natebosch** [comment]
> Bazel simply must support platform specific builds. It's certainly true that two builds for different platforms don't share compilation artifacts. Why do we care?

One issue with this is that we _use_ the Dart vm during builds for the web. Bazel level platform selection applies to the entire build - we'd be unable to build a tool and then us it with the VM during the build (for the web) if we rely on full-tree level platform selection. @jakemac53 - do you know whether there is support for build tools using a different build tree than the project? I would think there _should_ be...

My intuition is still that the bazel platform concept is not going to be a good fit but we should investigate it further. I think we're much more likely going to need to update our skylark rules to support the concept of multiple kernel files per target.

**Comment by leafpetersen** [comment]
> I think we're much more likely going to need to update our skylark rules to support the concept of multiple kernel files per target.

But this *would* slow down builds, right?  As opposed to parallel builds where any other platforms are off the critical path, this puts generating `n` platforms on the critical path, even though you probably almost never care about `n-1` of them.

**Comment by natebosch** [comment]
> But this _would_ slow down builds, right? As opposed to parallel builds where any other platforms are off the critical path, this puts generating `n` platforms on the critical path, even though you probably almost never care about `n-1` of them.

Insofar as having more actions makes builds slower yes this would be an issue. I don't think that all platforms would run serially in the critical path, and we might be able to use aspects to only build the platforms that we need.

**Comment by matanlurey** [comment]
@natebosch @jakemac53 Can we update this thread with the current status? I see we have some (lofty) bugs/features assigned, does that mean we are taking this up? What's the plan here?

**Comment by natebosch** [comment]
Current status from the build system side:

We're assuming we will need to be able to handle specialized kernel per-platform and we're starting work on those designs and scoping out the work. This assumption is looking likely.

We're tentatively assuming we will need to support config specific imports and we're also scoping out the work for that.

We are worried that our current strategy for always using the same SDK for codegen might cause problems and this is still an open discussion.

**Comment by dgrove** [comment]
@leafpetersen can you provide an update on what will be happening for Dart 2?

**Comment by natebosch** [comment]
From the build system side we've been busy with other Dart 2 urgent tasks and haven't had time to build a prototype - but we've had a few more internal discussions about this.

At this point I'm comfortable making the assumption that we'll be able to support the `dart.library.*` selectors and release with feature without a flag. We'll continue to consider the design alongside other changes (like platform specific dill files and modular kernel generation which we want outside of config specific imports) and raise a flag if we discover a showstopper but I think the risk is low.

**Comment by leafpetersen** [comment]
Sounds good.  So POR for Dart 2 is that we ship the feature as currently implemented in the kernel front end, without a flag.  I don't think there are any remaining blocking Dart 2 issues for this, but there are some tooling issues that should be improved.  I'll file some issues.

**Comment by dgrove** [comment]
Thanks - moving this out of the Dart2Stable milestone.

**Comment by leafpetersen** [comment]
Tracking issue for tooling support is here:  https://github.com/dart-lang/sdk/issues/33347 .

I'm going to close this out in favor of that issue, feel free to re-open if there is further discussion necessary.
